File: integration_tests/kotlin_client_tests/client-app/src/test/kotlin/dev/graphlink/test/QueryStringsTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.client.KotlinxSerializationGraphLinkJsonCodec
import dev.graphlink.test.generated.enums.UserStatus
import dev.graphlink.test.generated.inputs.AddressInput
import dev.graphlink.test.generated.inputs.CreateUserInput
import dev.graphlink.test.generated.inputs.UpdateUserInput
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.kPostJson
import dev.graphlink.test.Fixtures.kUserAliceJson
import dev.graphlink.test.Fixtures.newClient

/**
 * Regression tests for Kotlin string interpolation in generated query strings.
 *
 * GraphQL variable syntax uses `$varName` which is also Kotlin's string
 * interpolation syntax. The generator must emit `\$varName` so the literal
 * dollar sign survives at runtime. These tests decode the raw outgoing payload
 * and assert the `query` field contains `$` as a real character — not the
 * toString() of a Kotlin object substituted by the compiler.
 */
class QueryStringsTest {

    private val codec = KotlinxSerializationGraphLinkJsonCodec()
    private lateinit var adapter: MockAdapter
    private lateinit var client: GraphLinkClient

    private val kAddressInput = AddressInput("123 Main St", "Springfield", "US", "12345")
    private val kCreateInput = CreateUserInput(
        name = "Alice Smith",
        email = "alice@test.com",
        status = UserStatus.ACTIVE,
        address = kAddressInput,
    )

    @BeforeTest
    fun setUp() {
        adapter = MockAdapter()
        client = newClient(adapter, MockWebSocketAdapter())
    }

    private fun lastQuery(): String {
        val raw = adapter.getLastCall()!!.rawPayload
        return codec.decode(raw)["query"] as String
    }

    // ── Mutations ─────────────────────────────────────────────────────────────

    @Nested
    inner class MutationQueryStrings {

        @Test
        fun createUser_query_containsLiteralDollarInput() = runTest {
            adapter.registerData("createUser", mapOf("createUser" to kUserAliceJson))
            client.mutations.createUser(kCreateInput)
            val query = lastQuery()
            assertTrue(query.contains("\$input"),
                "Expected '\$input' in query but got: $query")
        }

        @Test
        fun updateUser_query_containsLiteralDollarId_andDollarInput() = runTest {
            adapter.registerData("updateUser", mapOf("updateUser" to kUserAliceJson))
            client.mutations.updateUser("user-1", UpdateUserInput(name = "Alice"))
            val query = lastQuery()
            assertTrue(query.contains("\$id"),
                "Expected '\$id' in query but got: $query")
            assertTrue(query.contains("\$input"),
                "Expected '\$input' in query but got: $query")
        }

        @Test
        fun deleteUser_query_containsLiteralDollarId() = runTest {
            adapter.registerData("deleteUser", mapOf("deleteUser" to true))
            client.mutations.deleteUser("user-1")
            val query = lastQuery()
            assertTrue(query.contains("\$id"),
                "Expected '\$id' in query but got: $query")
        }

        @Test
        fun transferPost_query_containsLiteralDollarArgs() = runTest {
            adapter.registerData("transferPost", mapOf("transferPost" to kPostJson))
            client.mutations.transferPost("post-1", "user-2")
            val query = lastQuery()
            assertTrue(query.contains("\$postId"),
                "Expected '\$postId' in query but got: $query")
            assertTrue(query.contains("\$newAuthorId"),
                "Expected '\$newAuthorId' in query but got: $query")
        }

        @Test
        fun query_doesNotContainInputObjectToString() = runTest {
            adapter.registerData("createUser", mapOf("createUser" to kUserAliceJson))
            client.mutations.createUser(kCreateInput)
            val query = lastQuery()
            assertFalse(query.contains("CreateUserInput("),
                "Query must not contain Kotlin object toString; got: $query")
        }
    }

    // ── Simple queries ────────────────────────────────────────────────────────

    @Nested
    inner class QueryQueryStrings {

        @Test
        fun getUser_query_containsLiteralDollarId() = runTest {
            adapter.registerData("getUser", mapOf("getUser" to kUserAliceJson))
            client.queries.getUser("user-1")
            val query = lastQuery()
            assertTrue(query.contains("\$id"),
                "Expected '\$id' in query but got: $query")
        }

        @Test
        fun listUsersByStatus_query_containsLiteralDollarStatus() = runTest {
            adapter.registerData("listUsersByStatus", mapOf("listUsersByStatus" to listOf(kUserAliceJson)))
            client.queries.listUsersByStatus(UserStatus.ACTIVE)
            val query = lastQuery()
            assertTrue(query.contains("\$status"),
                "Expected '\$status' in query but got: $query")
        }
    }

    // ── Multi-resolver (partial query) path ───────────────────────────────────

    @Nested
    inner class PartialQueryStrings {

        @Test
        fun fetchUserAndPost_partialQueryStrings_containLiteralDollar() = runTest {
            adapter.registerData("fetchUserAndPost", mapOf(
                "user" to kUserAliceJson, "post" to kPostJson,
            ))
            client.queries.fetchUserAndPost("user-1", "post-1")
            val query = lastQuery()
            assertTrue(query.contains("\$userId"),
                "Expected '\$userId' in partial query but got: $query")
            assertTrue(query.contains("\$postId"),
                "Expected '\$postId' in partial query but got: $query")
        }

        @Test
        fun fetchUserAndPost_query_doesNotContainObjectToString() = runTest {
            adapter.registerData("fetchUserAndPost", mapOf(
                "user" to kUserAliceJson, "post" to kPostJson,
            ))
            client.queries.fetchUserAndPost("user-1", "post-1")
            val query = lastQuery()
            assertFalse(query.contains("user-1") && query.contains("post-1") && !query.contains("\$userId"),
                "Query must not contain interpolated arg values; got: $query")
        }
    }
}

File: integration_tests/swift_client_tests/Tests/GraphLinkGeneratedTests/QueryStringsTest.swift
import XCTest
@testable import GraphLinkGenerated

/// Regression tests for literal `$varName` GraphQL variable syntax in
/// generated query strings.
///
/// Unlike Kotlin, Swift string interpolation uses `\(...)`, not `$name` — so
/// there's no compiler ambiguity to guard against here. These tests are
/// ported for parity anyway: they decode the raw outgoing payload and assert
/// the `query` field contains the literal `$varName` GraphQL variable
/// declarations the generator emitted.
private func lastQuery(_ adapter: MockAdapter) throws -> String {
    let raw = adapter.getLastCall()!.rawPayload
    let decoded = try GraphLinkJson.decode(raw.data(using: .utf8)!)
    return decoded["query"] as! String
}

private let kAddressInput = AddressInput(street: "123 Main St", city: "Springfield", country: "US", zip: "12345")
private let kCreateInput = CreateUserInput(
    name: "Alice Smith",
    email: "alice@test.com",
    status: .active,
    address: kAddressInput
)

// ── Mutations ────────────────────────────────────────────────────────────────

final class QueryStringsTest_MutationQueryStrings: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
    }

    func test_createUser_query_containsLiteralDollarInput() async throws {
        adapter.registerData("createUser", n(("createUser", Fixtures.kUserAliceJson)))
        _ = try await client.mutations.createUser(input: kCreateInput)
        let query = try lastQuery(adapter)
        XCTAssertTrue(query.contains("$input"), "Expected '$input' in query but got: \(query)")
    }

    func test_updateUser_query_containsLiteralDollarId_andDollarInput() async throws {
        adapter.registerData("updateUser", n(("updateUser", Fixtures.kUserAliceJson)))
        _ = try await client.mutations.updateUser(id: "user-1", input: UpdateUserInput(name: "Alice"))
        let query = try lastQuery(adapter)
        XCTAssertTrue(query.contains("$id"), "Expected '$id' in query but got: \(query)")
        XCTAssertTrue(query.contains("$input"), "Expected '$input' in query but got: \(query)")
    }

    func test_deleteUser_query_containsLiteralDollarId() async throws {
        adapter.registerData("deleteUser", n(("deleteUser", true)))
        _ = try await client.mutations.deleteUser(id: "user-1")
        let query = try lastQuery(adapter)
        XCTAssertTrue(query.contains("$id"), "Expected '$id' in query but got: \(query)")
    }

    func test_transferPost_query_containsLiteralDollarArgs() async throws {
        adapter.registerData("transferPost", n(("transferPost", Fixtures.kPostJson)))
        _ = try await client.mutations.transferPost(postId: "post-1", newAuthorId: "user-2")
        let query = try lastQuery(adapter)
        XCTAssertTrue(query.contains("$postId"), "Expected '$postId' in query but got: \(query)")
        XCTAssertTrue(query.contains("$newAuthorId"), "Expected '$newAuthorId' in query but got: \(query)")
    }

    func test_query_doesNotContainInputObjectToString() async throws {
        adapter.registerData("createUser", n(("createUser", Fixtures.kUserAliceJson)))
        _ = try await client.mutations.createUser(input: kCreateInput)
        let query = try lastQuery(adapter)
        XCTAssertFalse(query.contains("CreateUserInput("), "Query must not contain a stringified input object; got: \(query)")
    }
}

// ── Simple queries ───────────────────────────────────────────────────────────

final class QueryStringsTest_QueryQueryStrings: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
    }

    func test_getUser_query_containsLiteralDollarId() async throws {
        adapter.registerData("getUser", n(("getUser", Fixtures.kUserAliceJson)))
        _ = try await client.queries.getUser(id: "user-1")
        let query = try lastQuery(adapter)
        XCTAssertTrue(query.contains("$id"), "Expected '$id' in query but got: \(query)")
    }

    func test_listUsersByStatus_query_containsLiteralDollarStatus() async throws {
        adapter.registerData("listUsersByStatus", n(("listUsersByStatus", [Fixtures.kUserAliceJson])))
        _ = try await client.queries.listUsersByStatus(status: .active)
        let query = try lastQuery(adapter)
        XCTAssertTrue(query.contains("$status"), "Expected '$status' in query but got: \(query)")
    }
}

// ── Multi-resolver (combined) query path ────────────────────────────────────

final class QueryStringsTest_PartialQueryStrings: XCTestCase {
    var adapter: MockAdapter!
    var client: GraphLinkClient!

    override func setUp() {
        adapter = MockAdapter()
        client = Fixtures.newClient(adapter, MockWebSocketAdapter())
    }

    func test_fetchUserAndPost_partialQueryStrings_containLiteralDollar() async throws {
        adapter.registerData("fetchUserAndPost", n(
            ("user", Fixtures.kUserAliceJson), ("post", Fixtures.kPostJson)
        ))
        _ = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        let query = try lastQuery(adapter)
        XCTAssertTrue(query.contains("$userId"), "Expected '$userId' in query but got: \(query)")
        XCTAssertTrue(query.contains("$postId"), "Expected '$postId' in query but got: \(query)")
    }

    func test_fetchUserAndPost_query_doesNotContainObjectToString() async throws {
        adapter.registerData("fetchUserAndPost", n(
            ("user", Fixtures.kUserAliceJson), ("post", Fixtures.kPostJson)
        ))
        _ = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        let query = try lastQuery(adapter)
        let containsInterpolatedValues = query.contains("user-1") && query.contains("post-1") && !query.contains("$userId")
        XCTAssertFalse(containsInterpolatedValues, "Query must not contain interpolated arg values; got: \(query)")
    }
}
